Add query param to limit highlighting to specified length

matriv opened the pull request:
Add a `max_analyzed_offset` query parameter to allow users
to limit the highlighting of text fields to a value less than or equal to the
`index.highlight.max_analyzed_offset`, thus avoiding an exception when
the length of the text field exceeds the limit. The highlighting still takes place,
but stops at the length defined by the new parameter.

Closes: #52155

matriv committed:
Add option to limit highlighting to max offset

Add a query parameter `limit_to_max_analyzed_offset` to allow users
to limit the highlighting of text fields to the value of the
`index.highlight.max_analyzed_offset`, thus preventing from throwing
an exception when the length of the text field exceeds the limit.
The highlighting still takes place but up to the length set by the
setting.

Relates to: #52155

romseygeek commented on plugins/mapper-annotated-text/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighterTests.java:
```diff
@@ -59,78 +60,97 @@
+            for (int i = 0; i < snippets.length; i++) {
+                assertEquals(expectedPassages[i], snippets[i].getText());
+            }
+        } finally {
```
You can use try-with-resources here I think?

matriv replied:
Thought about it but we need to tries since, something like that: https://gist.github.com/matriv/ca6ca5fe191c8e0b1e49af282ec90669

so I preferred the traditional way, but happy to change.

romseygeek commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -206,6 +211,7 @@ public final void writeTo(StreamOutput out) throws IOException {
             out.writeMap(options);
         }
         out.writeOptionalBoolean(requireFieldMatch);
+        out.writeOptionalBoolean(limitToMaxAnalyzedOffset);
```
you'll need to do a version check here so that mixed-cluster highlighting works

matriv replied:
Done it already, thx.

romseygeek reviewed (commented):
Analyzer implementation LGTM, I'll defer to Jim and/or Mark on the API.  Does it make sense to have a separate query parameter, rather than doing everything with the index setting?  That is, would it be easier to use if we just changed things so that you always get the limiting filter added, and you don't get exceptions any more?

matriv committed:
fix bwc

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

matriv committed:
fix version check for bwc

matriv committed:
skip for previous versions

jimczi commented:
> , I'll defer to Jim and/or Mark on the API. Does it make sense to have a separate query parameter, rather than doing everything with the index setting? That is, would it be easier to use if we just changed things so that you always get the limiting filter added, and you don't get exceptions any more?

+1 for a separate query parameter. I'd also like to remove that setting but not sure if we need to do it in the same PR. Maybe we can consider this new option independently of the fact that the setting to limit highlighting is not useful ?

matriv commented:
> > , I'll defer to Jim and/or Mark on the API. Does it make sense to have a separate query parameter, rather than doing everything with the index setting? That is, would it be easier to use if we just changed things so that you always get the limiting filter added, and you don't get exceptions any more?
> 
> +1 for a separate query parameter. I'd also like to remove that setting but not sure if we need to do it in the same PR. Maybe we can consider this new option independently of the fact that the setting to limit highlighting is not useful ?

Sorry, I'm a bit confused @jimczi . We should add this query param as is already in this PR, correct?
I don't get the rest, could you please explain a bit more about the `the setting to limit highlighting is not useful` ?

jimczi commented:
Can you describe what's implemented with an example ? What's the relation between the query parameter and the setting ? Do we need both ? My +1 was for a dedicated query parameter that is independent of the setting.

markharwood commented:
>would it be easier to use if we just changed things so that you always get the limiting filter added, and you don't get exceptions any more?

Taking the perspectives of the various parties involved and what they want:
1) The es cluster admin wants to shut down abusive apps with noisy failures (the point of the existing index setting). 
2) The client app developer has a new setting to write a more responsible app which doesn't blow up.
3) The end user doing a search on big docs no longer gets errors if the app developer has used the flag but doesn't get a "truncated" warning.


The existing error message below for exceeding the `index.highlight.max_analyzed_offset` is not ideal.
```
 "The length of [text] field of [0] doc of [index] index has exceeded [xx] - maximum allowed to be analyzed for "
                + "highlighting. This maximum can be set by changing the [index.highlight.max_analyzed_offset] index level setting. "
                + "For large texts, indexing with offsets or term vectors is recommended!",
```
It should probably offer an alternative remedy. This is the opportunity for the admin to tell the client app developer to write a better app so we should advertise the new `limit_to_max_analyzed_offset ` flag in this message rather than only suggesting upping the `index.highlight.max_analyzed_offset` limit.

With regards to point 3 - the client app developer might want to tell users when content is truncated. This could be done by testing if the length of the highlighted string is greater than the `limit_to_max_analyzed_offset` setting but this is not a reliable test. A truncated flag in the response would be more robust but maybe warning users that text has been truncated is not too important given the hit-or-miss nature of highlighters anyway.

matriv commented:
@markharwood I had in mind to have a follow up PR to have this warning/flag regarding the "truncation" of the highlighting, since I agree this is useful to avoid confusion for the user. For "simple" users, I guess they will first try to highlight a large doc, get the error, and then use the flag to highlight up to the limit, so they are aware of what's going on, therefore, of course I need to add the new param to the error message as a suggestion.

As you mentioned in point 1. seem we want to have both setting on the index and the per-query param.

Maybe, if we have a query param that has a similar name with the setting like: `max_analyzed_offset` which overrides the index setting and uses the filter to stop highlighting and avoid exceptions? (@jimczi probably what you're suggesting?). To me this doesn't sound bad, but I think we need to consider that in this case, the query param doesn't only override the limit of the index setting, but also changes the behaviour (query now succeeds with truncated highlighting).

matriv committed:
Add suggestion to use new query param to the exception message

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
reword error message and adjust tests

matriv commented:
Reworded the expression with the help of @romseygeek, and removed the mention to index offsets and term vectors.

matriv committed:
change wrapping method, fix AnnotatedTextHighlighter

matriv committed:
remove uneccessary override of method

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
remove unused import

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
revert remove mode

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

mark-vieira commented:
@elasticmachine update branch

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
fix licence

matriv committed:
Change semantics of the new setting.

It's converted to an Integer setting that doesn't override the index's
highlight.max_analyzed_offset setting.

matriv committed:
revert import order change

matriv committed:
fix checkstyle

matriv committed:
fix checkstyle

matriv committed:
add more tests

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
fix checkstyle

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv commented:
Changed the name and semantics of the new param. It's accepts an integer value for the max offset and doesn't override the index setting which is checked independently. So the user needs to specify the query param with a value lower than the index setting value, to highlight a portion of the text and avoid the exceptions.

romseygeek reviewed (approved):
LGTM, thanks for the iterations @matriv

markharwood commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java:
```diff
@@ -89,23 +90,26 @@ public HighlightField highlight(FieldHighlightContext fieldContext) throws IOExc
+                    "The length [" + textLength + "] of field [" + field +"] in doc[" + hitContext.hit().getId() + "]/index["
+                        + context.getIndexName() +"] exceeds the [" + IndexSettings.MAX_ANALYZED_OFFSET_SETTING.getKey() + "] "
+                        + "limit [" + maxAnalyzedOffset + "]. To ignore text beyond this limit when highlighting, set the query "
+                        + "parameter [" + MAX_ANALYZED_OFFSET_FIELD.toString() + "] to the desired value."
```
I'm not sure "Set the parameter ... to the desired value" gives users enough of a clue to the solution. It should say something about setting the value lower than the maxAnalyzedOffset index value. eg:
> "By setting the query parameter X to a value lower than the index limit Y, long fields will be returned truncated rather than causing this error"

matriv committed:
address comments

matriv committed:
Merge remote-tracking branch 'upstream/master' into limitHighlighting

matriv committed:
rephrase error message
